gb-dma/src/hdma.rs

```diff
@@ -0,0 +1,162 @@
+use gb_bus::{Address, Bus, Error, FileOperation, IORegArea, Lock};
+use gb_clock::{Tick, Ticker};
+use gb_cpu::cpu::Cpu;
+use gb_ppu::{drawing, Ppu};
+
+#[derive(PartialEq)]
+pub enum HdmaMode {
+    Gdma,
+    Hdma,
+}
+
+#[derive(Default)]
+pub struct Hdma {
+    src: u16,
+    dest: u16,
+    active: bool,
+    data_chunks_len: u8,
+    current_chunk_len: u8,
+    last_ppu_mode: Option<drawing::Mode>,
+    mode: Option<HdmaMode>,
+}
+
+impl Hdma {
+    const DEST_STARTING_ADDR: u16 = 0x8000;
+    const DATA_CHUNK_SIZE: u8 = 0x10;
+    const MAX_DATA_CHUNKS_LEN: u8 = 0x7F;
+    const HDMA_MODE_BIT: u8 = 0x80;
+    const BYTES_PER_CYCLE: u8 = 2;
+
+    pub fn new_data_chunk(&mut self) {
+        self.current_chunk_len = Self::DATA_CHUNK_SIZE;
+    }
+
+    fn data_transfer(&mut self, adr_bus: &mut dyn Bus<u8>) {
+        let v = adr_bus
+            .read(self.src, Some(Lock::Dma))
+            .expect("memory unavailable during HDMA");
+        if adr_bus.write(self.dest, v, Some(Lock::Dma)).is_err() {
+            log::error!(
+                "failed to write data '{:x}' at '{:x}' during HDMA",
+                v,
+                self.dest
+            );
+        }
+        self.src += 1;
+        self.dest += 1;
+    }
+
+    // Method used before each machine cycle to check hdma status
+    pub fn check_hdma_state(&mut self, mut cpu: &mut Cpu, ppu: &Ppu) {
+        if self.active {
+            cpu.halted = match self.mode {
+                Some(HdmaMode::Gdma) => {
```

Comment:
for reference [General Purpose DMA](https://gbdev.io/pandocs/CGB_Registers.html#bit-7--0---general-purpose-dma) (for other to read)